File: tours-service/service/tour_execution_service.go
package service

import (
    "errors"
    "time"
    "tours-service/model"
    "tours-service/repo"

    "go.mongodb.org/mongo-driver/bson/primitive"
)

// Create a new TourExecution
// service
func CreateTourExecution(tourID, touristID string, lat, lng float64) (model.TourExecution, error) {
    tourOID, err := primitive.ObjectIDFromHex(tourID)
    if err != nil { return model.TourExecution{}, errors.New("neispravan tour ID") }
    touristOID, err := primitive.ObjectIDFromHex(touristID)
    if err != nil { return model.TourExecution{}, errors.New("neispravan tourist ID") }

    now := time.Now().UTC()
    exec := model.TourExecution{
        TourID:         tourOID,
        TouristID:      touristOID,
        Status:         model.Active,
        StartedAt:      primitive.NewDateTimeFromTime(now),
        LastActivityAt: primitive.NewDateTimeFromTime(now),
        KeyPoints:      []model.KeyPointExecution{},
    }

    if err := repo.AddTourExecution(&exec); err != nil {
        return model.TourExecution{}, err
    }
    return exec, nil // exec.ID is now set
}

// Get TourExecution by ID
func GetTourExecutionByID(executionID primitive.ObjectID) (model.TourExecution, error) {
    return repo.GetTourExecutionByID(executionID)
}

// Change status of a TourExecution (completed or abandoned)
func UpdateTourExecutionStatus(executionID string, status model.TourExecutionStatus) error {
    execObjectID, err := primitive.ObjectIDFromHex(executionID)
    if err != nil {
        return errors.New("neispravan execution ID")
    }
    now := primitive.NewDateTimeFromTime(time.Now())
    return repo.UpdateTourExecutionStatus(execObjectID, status, now)
}

// Add completed key point to TourExecution
func AddCompletedKeyPoint(executionID string, keyPointID string) error {
    execObjectID, err := primitive.ObjectIDFromHex(executionID)
    if err != nil {
        return errors.New("neispravan execution ID")
    }
    kpObjectID, err := primitive.ObjectIDFromHex(keyPointID)
    if err != nil {
        return errors.New("neispravan key point ID")
    }
    now := primitive.NewDateTimeFromTime(time.Now())
    return repo.AddKeyPointToExecution(execObjectID, kpObjectID, now)
}

File: frontend/src/services/tourExecutionService.ts
import api from '../api';
import { Location } from '../types/user';

// Interface for the tour execution response
export interface TourExecutionResponse {
  id: string;
  tourId: string;
  touristId: string;
  status: string;
  startedAt: string;
  endedAt?: string;
  lastActivityAt: string;
  keyPoints: Array<{
    keyPointId: string;
    completedAt: string;
    name?: string;
  }>;
}

export const tourExecutionService = {
  // Start a new tour execution
  createTourExecution: async (tourId: string, location: Location): Promise<TourExecutionResponse> => {
    const response = await api.post(`/api/tours-service/tours/${tourId}/execute`, { location });
    return response.data as TourExecutionResponse;
  },

  // Get tour execution details by executionId
  getTourExecutionById: async (executionId: string): Promise<TourExecutionResponse> => {
    const response = await api.get(`/api/tours-service/tour-executions/${executionId}`);
    return response.data as TourExecutionResponse;
  },

  // Update tour execution status (complete/abandon)
  updateTourExecutionStatus: async (executionId: string, status: string) => {
    const response = await api.put(`/api/tours-service/tour-executions/${executionId}/status`, { status });
    return response.data;
  },

  // Add completed key point to tour execution
  addCompletedKeyPoint: async (executionId: string, keyPointId: string) => {
    const response = await api.post(`/api/tours-service/tour-executions/${executionId}/keypoint`, { keyPointId });
    return response.data;
  }
};